libbeat/cmd/instance/locks/lock.go

```diff
@@ -30,233 +28,81 @@ import (
 	return &Locker{
-		fileLock:  flock.New(lockfilePath),
-		logger:    logp.L(),
-		beatName:  beatInfo.Beat,
-		filePath:  lockfilePath,
-		beatStart: beatInfo.StartTime,
+		fileLock:   flock.New(lockfilePath),
+		retryCount: retryCount,
+		retrySleep: retrySleep,
+		logger:     logp.L(),
 	}
 }
 
 // Lock attempts to acquire a lock on the data path for the currently-running
 // Beat instance. If another Beats instance already has a lock on the same data path
 // an ErrAlreadyLocked error is returned.
 func (lock *Locker) Lock() error {
-	new := pidfile{PID: pidFetch(), WriteTime: time.Now()}
-	encoded, err := json.Marshal(&new)
-	if err != nil {
-		return fmt.Errorf("error encoding json for pidfile: %w", err)
-	}
-
-	// The combination of O_CREATE and O_EXCL will ensure we return an error if we don't
-	// manage to create the file
-	fh, openErr := os.OpenFile(lock.filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o600)
-	// Don't trust different OSes to report the errors we expect, just try to recover regardless
-	if openErr != nil {
-		err = lock.handleFailedCreate()
+	for i := 0; i < lock.retryCount; i++ {
+		gotLock, err := lock.fileLock.TryLock()
 		if err != nil {
-			return fmt.Errorf("cannot obtain lockfile: %w", err)
+			return fmt.Errorf("unable to try a lock of the data path: %w", err)
 		}
-		// If something fails here, it's probably unrecoverable
-		fh, err = os.OpenFile(lock.filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o600)
-		if err != nil {
-			return fmt.Errorf("cannot re-obtain lockfile %s: %w", lock.filePath, err)
+		if gotLock {
+			return nil
 		}
+		lock.logger.Debugf("Could not obtain lock for file %s, retrying %d times", lock.fileLock.Path(), (lock.retryCount - i))
+		time.Sleep(lock.retrySleep)
 	}
-
-	// a Process can't write to its own locked file on all platforms, write first
-	_, err = fh.Write(encoded)
-	if err != nil {
-		return fmt.Errorf("error writing pidfile to %s: %w", lock.filePath, err)
-	}
-
-	// Exclusive lock
-	isLocked, err := lock.fileLock.TryLock()
-	if err != nil {
-		return fmt.Errorf("unable to lock data path: %w", err)
-	}
-	// case: lock could not be obtained.
-	if !isLocked {
-		// if we're here, things are probably unrecoverable, as we've previously checked for a lockfile. Exit.
-		return fmt.Errorf("%s: %w", lock.filePath, ErrAlreadyLocked)
-	}
-
-	return nil
+	return fmt.Errorf("%s: %w", lock.fileLock.Path(), ErrAlreadyLocked)
 }
 
 // Unlock attempts to release the lock on a data path previously acquired via Lock().
 func (lock *Locker) Unlock() error {
-	err := lock.fileLock.Unlock()
-	if err != nil {
-		return fmt.Errorf("unable to unlock data path: %w", err)
-	}
-
-	err = os.Remove(lock.fileLock.Path())
-	if err != nil {
-		return fmt.Errorf("unable to unlock data path file %s: %w", lock.fileLock.Path(), err)
-	}
-	return nil
-}
+	// remove first while we still have the lock, so we reduce the odds of another beat swooping in to start between the Unlock() and Remove() operation.
```

Comment:
I would specifically mention that `TryLock` doesn't set the `os.O_EXCL`, I think that is an important detail from the comment chain in https://github.com/elastic/beats/pull/34194/files#r1064871207